_posts/2018-03-15-complete-introduction-to-nodejs.md

```diff
@@ -0,0 +1,406 @@
+```
+
+When you've watched the video about how the V8 Engine works, you'll know why 'hello' get's logged first and 'world' second:
+What happens is:
+- setTimeout is added to the call stack,
+- setTimeout has a timer and a callback, this fire's up  V8's timer webapi,
+- now that the webapi is taking care of the setTimeout, it get's removed from the call stack,
+- console.log is added to the call stack, it logs hello, and removes console.log from the call stack,
+- once the timer has completed, it pushes the callback to the task queue,
+- Since there are no more function calls on the call stack, the event loop adds the callback to the call stack
+- 'world' is printed.
+
+There's no need for an additional thread to pause the program for 2 seconds and after that log 'world',
+the V8 Engine handles this for us just like it does with any other async functionality in the browser.
+So this is a very simple example of how non-blocking code works in NodeJs, the timeout didn't block our code, 'hello' got logged right away.
+
+## Hello Module
+Now lets give you an example how to create a node module (a very simple and not very useful one) but just to give you an idea of how you can export your code and use it in other files.
+
+We're going to create a module that has a log function which takes a parameter (name) and logs 'hello <name>'
+to the console. Then we'll import it in another file and call the function from there.
+
+`log.js`
+
+ ```javascript
+// Create our custom hello function
+const hello = function(name){ console.log(`Hello ${name}`); } 
+// export this functionality 
+module.exports = log;
+```
+
+`hello.js`
+
+```javascript
+// the way we import another module into our file is by using require(), require is a global module for node
+// when requiring local modules (not the ones we install with npm), we give it the path relative to the current file,
+// no need for extensions, since node looks for a .js file
+var log = require('./log');
+
+// now that we have our functionality available in the log variable, we can use it
+log('Mike'); 
+// log's 'hello Mike' to the console
+```
+
+To import modules that you've installed with npm, don't specify a path, but give it the package's name.
+For example
+```
+// this imports the express module from the express package if we installed it`
+require('express');
+```
+
+To learn more about how to use require go check out this useful url: [https://medium.freecodecamp.org/requiring-modules-in-node-js-everything-you-need-to-know-e7fbd119be8](https://medium.freecodecamp.org/requiring-modules-in-node-js-everything-you-need-to-know-e7fbd119be8)
+
+## Hello Server
+Now that we know how to require other modules, let's create a basic server application.
+ We're not going to install any server frameworks (like express) yet, instead we'll require a module that come with node.
+Node has some built in modules that you can use like http (http server), https, fs (file system), net (tcp sockets), … <br>
```

Comment:
built-in

If they are commands, wrap them in backticks:
`http`
`https`
`fs`
`net`

Reply:
no commands, modules that can be required in files, but probably best in backticks as well